### ISSUE TITLE
Static linking on MSVC

### ISSUE THREAD

**Issue by Swat-SomeBug** [libsigcplusplus/libsigcplusplus#90]
Hi,
When [packaging](https://github.com/build2-packaging/libsigcplusplus.git) the library for build2, I ran into an issue with msvc and static linking.
Based on [this](https://github.com/build2-packaging/libsigcplusplus/blob/2f34490d2850183b9db6bcae5dcaa7b580fdab35/libsigc%2B%2B/sigc%2B%2Bconfig.h.in#LL59C1-L69C23) code snippet, it seems, only dynamic linking is supported with msvc.
Is this true?
Currently, to support static linking on windows, `SIGC_BUILD` needs to be passed to consumers of the library while `_WINDLL` should not be passed, which is a way to get `SIGC_API` to be empty. This seems counter intuitive and occurs because `SIGC_DLL` is always set to `1` with msvc.
```cpp
#ifdef _WIN32
#if defined(_MSC_VER)
#define SIGC_MSC 1
#define SIGC_WIN32 1
#define SIGC_DLL 1
...
#ifdef SIGC_DLL
#if defined(SIGC_BUILD) && defined(_WINDLL)
#define SIGC_API __declspec(dllexport)
#elif !defined(SIGC_BUILD)
#define SIGC_API __declspec(dllimport)
#else
#define SIGC_API
#endif
#else /* !SIGC_DLL */
#define SIGC_API
#endif /* !SIGC_DLL */
```

To support both dynamic and static linking, could passing `SIGC_DLL` on the command line, be considered. The code snippet would look as below:
```cpp
#ifdef SIGC_DLL
#ifdef SIGC_BUILD
#define SIGC_API __declspec(dllexport)
#else
#define SIGC_API __declspec(dllimport)
#endif
#else /* !SIGC_DLL */
#define SIGC_API
#endif /* !SIGC_DLL */
```
Basically the snippet above requires the user to pass `-DSIGC_BUILD` and `-DSIGC_DLL` when building a shared library under windows and only `-DSIGC_DLL` when consuming the library as a DLL. `_WINDLL` is unnecessary in this case. For all other cases (static lib on windows or static/dynamic on any other platform), no defines are needed.

**Comment by kjellahl** [comment]
@fanc999 Can you have a look at this issue, please.

It looks like a static libsigc++ library can't be built with MSVC,
but a static libxml++ library can be built.

If you make it possible to build a static libsigc++ library, the type of library
shall of course be selected with the `default_library` meson option.

**Comment by fanc999-1** [comment]
Hi,

First looks, the historic thing was that we never really supported static builds of the gtkmm stack, so let's see whether we can make it happen here.  I'll try to see whether we can be building with `default_library=both`, but most probably that will be later, if it is feasible.

Sorry about the delay-I lost access to my original account and had to unlink my e-mail from it.

With blessings, and cheerws!

**Comment by kjellahl** [comment]
Fixed by PR #91 and #92.
